### PR TITLE
fixing transaction tests

### PR DESCRIPTION
**Pull request by daprahamian**
## Description

A few things:
- Fixed a bug where session was not getting passed to serverSelection, which broke mongos-pinning for transactions
- Fixed the error that is thrown for invalid sharded transaction sessions. Now that we support
sharded transactions, it should not throw on sharded unless wire version is low enough
- Fixed a bunch of test runner issues that were causing test failures
- Temporarily disabled some tests that could not be worked around in the current state of the unified test runner.

**Are there any files to ignore?**
No

### CODE REVIEW

File: lib/core/sessions.js [hunk 1]
```diff
@@ -670,6 +672,12 @@ function applySession(session, command, options) {
   const isRetryableWrite = options.willRetryWrite;
   const shouldApplyReadConcern = commandSupportsReadConcern(command);
 
+  // console.trace({
```
Comment (rosemaryy) [inline lib/core/sessions.js]: (nit) Remove these comments?

File: lib/core/utils.js [hunk 2]
```diff
@@ -87,6 +87,13 @@ function maxWireVersion(topologyOrServer) {
     return topologyOrServer.ismaster.maxWireVersion;
   }
 
+  if (typeof topologyOrServer.lastIsMaster === 'function') {
```
Comment (kvwalker) [inline lib/core/utils.js]: What `topologyOrServer` does not have the `lastIsMaster` function?

File: test/environments.js [hunk 3]
```diff
@@ -137,10 +137,12 @@ class ShardedEnvironment extends EnvironmentBase {
     this.host = 'localhost';
     this.port = 51000;
 
+    // Disabled only connecting to a single shard for now
```
Comment (kvwalker) [inline test/environments.js]: Can you add a brief description of why or a `TODO`?

File: test/functional/transactions_tests.js [hunk 4]
```diff
@@ -71,10 +76,10 @@ describe('Transactions', function() {
 
   describe('startTransaction', function() {
     it('should error if transactions are not supported', {
-      metadata: { requires: { topology: ['sharded'], mongodb: '>4.0.0' } },
+      metadata: { requires: { topology: ['sharded'], mongodb: '4.0.x' } },
       test: function(done) {
         const configuration = this.configuration;
-        const client = configuration.newClient(configuration.writeConcernMax(), { poolSize: 1 });
+        const client = configuration.newClient(configuration.url());
```
Comment (kvwalker) [inline test/functional/transactions_tests.js]: Do we not need to worry about `writeConcern` here?

File: lib/core/utils.js (hunk 2 above)
Comment (daprahamian) [reply lib/core/utils.js]: Technically `lib/core/sdam/server.js` does not have it. I don't think it is ever actually used in this function, but I wanted to be safe.

File: test/environments.js (hunk 3 above)
Comment (daprahamian) [reply test/environments.js]: will do

File: test/functional/transactions_tests.js (hunk 4 above)
Comment (daprahamian) [reply test/functional/transactions_tests.js]: In this specific case no, b/c the test should immediately error.

**Comment by daprahamian** [comment]
Updated commit messages and added a TODO describing why we went back to 2 shards in the environments file

**Pull request merged**